homeassistant/helpers/entity_component.py

```diff
@@ -229,9 +229,23 @@ async def async_setup_platform(
             return
 
         # Use config scan interval, fallback to platform if none set
-        scan_interval = platform_config.get(
-            CONF_SCAN_INTERVAL, getattr(platform, "SCAN_INTERVAL", None)
-        )
+        # FIXME: Components don't pass discovery_info in the correct format:
```

Comment:
@MartinHjelmare, @janiversen When we setup a platform through dynamic loading, current code ignores `scan_interval` config from `discovery_info`. Following code fixes that and now it works as expected.

Caveats:
- Platform code expects time interval to be defined as `datetime.timedelta(seconds=...)`. I can rework my code to use integer in my structure, however, in such a case we should do a conversion to timedelta in this place. Validation method `cv.time_period` does the conversion for us. The downside is `Recorder` warning, as mentioned.
```bash
2020-07-06 14:09:14 WARNING (Recorder) [homeassistant.components.recorder] Event is not JSON serializable: <Event platform_discovered[L]: service=load_platform.cover, platform=modbus, discovered=[OrderedDict([('name', 'plc_01_door'), ('hub', 'plc_01'), ('slave', 1), ('device_class', 'door'), ('scan_interval', datetime.timedelta(seconds=5)), ('coil', 0), ('status_register', 4), ('status_register_type', 'input'), ('state_opening', 1), ('state_open', 2), ('state_closing', 3), ('state_closed', 4)])]>
```
- When components set up platform, they pass `discovery_info` in different formats. Most components pass `None` when not initialized dynamically, however, the `Updater` component passes `{}`. Also, `cover` component loaded dynamically passes a list of a single dictionary, instead of passing the dictionary directly (expected `Optional[DiscoveryInfoType]`, which is `Dict[str, Any]`)
- I updated the code to make it work. However, the format of `DiscoveryInfoType` should be unified across the whole codebase. Temporarily, I changed the type to `Optional[Any]` to supress linter complaint.

Here's my debugging log. First line is the name of the component, second line is the content of `discovery_info`.
```bash
[vzahradnik@Ryzen home-assistant-core (feature/modbus-cover)]$ hass
homeassistant
None
updater
{}
modbus
None
modbus
None
modbus
None
modbus
[OrderedDict([('name', 'plc_01_door'), ('hub', 'plc_01'), ('slave', 1), ('device_class', 'door'), ('scan_interval', datetime.timedelta(seconds=5)), ('coil', 0), ('status_register', 4), ('status_register_type', 'input'), ('state_opening', 1), ('state_open', 2), ('state_closing', 3), ('state_closed', 4)])]
```